### PR TITLE
reflection and macro cleanup

### PR DESCRIPTION
**Pull request by xeno-by**
Scaladoc-driven cleanup rocks. Yesterday night the API was quite ugly, now it looks much better.

### CODE REVIEW

**Comment by xeno-by** [comment]
review @paulp @retronym @gkossakowski

**Comment by xeno-by** [comment]
also review @VladUreche for scaladoc commits

File: src/reflect/scala/reflect/api/Position.scala [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+ *  INV1: A tree with an offset position never contains a child
+ *        with a range position
+ *  INV2: If the child of a tree with a range position also has a range position,
+ *        then the child's range is contained in the parent's range.
+ *  INV3: Opaque range positions of children of the same node are non-overlapping
+ *        (this means their overlap is at most a single point).
+ *
+ *  The following tests are useful on positions:
+ *
+ *  pos.isDefined     true if position is not a NoPosition nor a FakePosition
```
Comment (VladUreche) [inline src/reflect/scala/reflect/api/Position.scala:22]: You can enclose these in a {{{code block}}} to get a monospaced font.

File: src/reflect/scala/reflect/api/Symbols.scala [hunk 2]
```diff
@@ -86,10 +86,10 @@ trait Symbols { self: Universe =>
      *  that directly contains the current symbol's definition.
      *  The `NoSymbol` symbol does not have an owner, and calling this method
      *  on one causes an internal error.
-     *  The owner of the Scala root class [[scala.reflect.api.MirrorOf.RootClass]]
-     *  and the Scala root object [[scala.reflect.api.MirrorOf.RootPackage]] is `NoSymbol`.
+     *  The owner of the Scala root class [[scala.reflect.api.Mirror.RootClass]]
+     *  and the Scala root object [[scala.reflect.api.Mirror.RootPackage]] is `NoSymbol`.
      *  Every other symbol has a chain of owners that ends in
-     *  [[scala.reflect.api.MirrorOf.RootClass]].
+     *  [[scala.reflect.api.Mirror.RootClass]].
```
Comment (VladUreche) [inline src/reflect/scala/reflect/api/Symbols.scala:92]: You can also use relative links, if it's obvious to what they point to.

File: src/reflect/scala/reflect/api/Position.scala (hunk 1 above)
Comment (xeno-by) [inline src/reflect/scala/reflect/api/Position.scala:22]: This PR intentionally doesn't touch the docs themselves. We'll have a follow-up next week. I'll add your suggestion to the doc meeting google doc. Thanks!

**Comment by VladUreche** [comment]
LGTM, I added two comments, but it's up to you if you want to act on them or not.

**Comment by gkossakowski** [comment]
Wow. That's huge. Do we really want all of that to be included? Also tests are failing.

**Comment by gkossakowski** [comment]
Given that tests are failing I'd like to drop this for 2.10.0. If you have some essential part of this PR that absolutely has to be in then let's discuss it.

**Comment by xeno-by** [comment]
PLS REBUILD ALL

**Comment by xeno-by** [comment]
The tests were failing, because I forgot to rename a method. It's no Friday yet, why close? This PR is directly affiliated with scaladoc improvements (reduces the surface for documentation, provides better names for top-level traits)

**Comment by xeno-by** [comment]
Regarding the size, well, each commit is a simple rename or removal, without any mechanisms changed.

File: src/reflect/scala/reflect/api/Position.scala (hunk 1 above)
Comment (retronym) [inline src/reflect/scala/reflect/api/Position.scala:22]: In fact, they could be links to the method docs. `[[Position.isDefined]]`

**Comment by gkossakowski** [comment]
We discussed this @jsuereth and it won't go into 2.10.0. It's too big change (even if it's only about moving code around, it also matters and introduced problems to the whole ecosystem we need to sync with). Also, it cannot be reviewed in reasonable time.

If there's anything that's absolutely critical from this long list of commits let's discuss it. The whole PR won't go in.

**Comment by odersky** [comment]
Chris and I just went through it with Eugene. I think each commit makes a lot of sense. All of it is cruft that we are much better off without. Removing this stuff is much more helpful than trying to document it, and we _have_ classified documentation as a blocker.  

I also like the effort to have more consistency in naming. 

So I would put a vote in to accept this. 233 net deletions is a big win. It will improve the quality of what we ship a lot.
Sure it's late, but it will be crucial to send a message that we care about API quality.

**Comment by jsuereth** [comment]
Martin -  

This is more about policy.  YES - We care about API quality.   However, we have to weight all things in context.   Reflection is currently something that can drastically affect other components, like the IDE.   At this point, if you touch a lot of code, there's higher risk we have undiscovered blockers show up in an RC1.  I'd rather _know_ we have a problem and fix it in 2.10.1 which will release four months later. 

We specifically made reflection experimental because we knew we wouldn't be ready for another 4 months.   We have to draw a line in the sand somewhere, well, we drew the line for 2.10.0.   If this pull request makes it into 2.10.0 we'll have to _UNDO_ a lot of work we did to draw the line.   It also means delaying the release by another month or so just to make sure everything is stable.

It's risk vs. reward right now.   Yes the API is important, important enough to release a 2.10.1 _even sooner_ than 2 months.   That's where this pull request belongs right now.

I understand that a few of the items touched were introduced in M7, however we already have people depending on these things for 2.10.0.  Momentum is against us and it takes a while for the pendulum to stop swinging after any "major refactorings".   Unfortunately, this is such a change.

That said, I'd like to leave this pull request up against 2.10.x.   We'll be moving 2.10.0 to the WIP branch shortly and then this can immediately go into 2.10.1 where we have ~2 months to think through changes like this before trying to harden things and stop the wild gyrations of change for the community.

Basically:  I'd rather release with known issues than unknown issues.

**Comment by xeno-by** [comment]
If we strictly follow the policy, then this PR doesn't violate it. It is a "cleanup and polish to _new features_ introduced in 2.10.x ", was submitted before Wed 12:00, got an LGTM before Friday midnight.

Also I'm very worried about the first impression people get from the reflection API (which is something that cannot be undone in 2.10.1). This PR is a direct result of me compiling a scaladoc and taking a cursory look. Unfortunately, that look revealed a lot of ugliness, which I did my best to fix.

Could you also please elaborate on "it also means delaying the release by another month or so just to make sure everything is stable"?

**Comment by gkossakowski** [comment]
We reached conclusion that @xeno-by will test sbt, Akka and IDE locally to see if it breaks. Josh will check if community build breaks. If not, we'll consider this to be purely touching reflection so if new problems arise by merging it they will be localized in reflection that is experimental and we can release it as-is.

If this affect a non-experimental area of the compiler, we'll revert for 2.10.0 and re-introduce in 2.10.1.

**Comment by xeno-by** [comment]
Rebased against the fresh trunk, because a recent commit changed the code that's been removed by this pull request.

**Comment by gkossakowski** [comment]
PLS REBUILD ALL

**Comment by jsuereth** [comment]
Kitteh got stuck on second build, it passed in the output.  This is good to go.
